### PR TITLE
Faucet

### PR DESCRIPTION
**Pull request by MrLuit**
Story details: https://app.clubhouse.io/mycrypto/story/1130

### CODE REVIEW

File: src/features/Faucet/Faucet.tsx [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+        } else {
+          setLoading(false);
+          setError(e.message);
+        }
+      });
+  };
+
+  const validAccounts = accounts.filter((account) => faucetNetworks.includes(account.network.name));
+
+  const txConfig = (() => {
```
Comment (FrederikBolding) [inline src/features/Faucet/Faucet.tsx:167]: Can we abstract this out of the component?

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

**Comment by FrederikBolding** [comment]
Please rebase too!

**Comment by FrederikBolding** [comment]
We need to wait for #3564 to test this properly. I rebased it to fix conflicts

File: src/features/Faucet/Faucet.tsx [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+`;
+
+const SubmitCaptchaButton = styled(Button)`
+  width: 100%;
+  margin-top: 30px;
+`;
+
+const faucetNetworks = ['Ropsten', 'Rinkeby', 'Kovan', 'Goerli'];
+
+export function Faucet({ history }: RouteComponentProps<{}>) {
```
Comment (FrederikBolding) [inline src/features/Faucet/Faucet.tsx:102]: Instead of exporting with `withRouter`, you can use `useHistory` inside the component

File: src/components/AccountList.tsx [hunk 3]
```diff
@@ -529,7 +529,7 @@ const BuildAccountTable = (
           </LabelWithWallet>
         </Label>,
         <EthAddress key={index} address={account.address} truncate={true} isCopyable={copyable} />,
-        <Network key={index} color="#a682ff">
+        <Network key={index} color={account.network.color || '#a682ff'}>
```
Comment (GuillaumeRx) [inline src/components/AccountList.tsx:532]: Use `COLORS` from `@theme` to define colors

File: src/features/Faucet/Faucet.tsx [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+const LearnMoreArrow = styled(Icon)`
+  transform: rotate(180deg);
+  width: 14px;
+  height: 14px;
+`;
+
+const NoTestnetAccounts = styled(InlineMessage)`
+  margin-bottom: 20px;
+  a {
+    color: var(--color-link-color);
```
Comment (GuillaumeRx) [inline src/features/Faucet/Faucet.tsx:70]: Use `COLORS` from `@theme` to define colors. SCSS shouldn’t be used since we want to get rid of it

File: src/features/Faucet/Faucet.tsx [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    text-decoration: none;
+  }
+  a:hover {
+    color: var(--color-link-color);
```
Comment (GuillaumeRx) [inline src/features/Faucet/Faucet.tsx:74]: Use `COLORS` from `@theme` to define colors. SCSS shouldn’t be used since we want to get rid of it

File: src/features/Faucet/Faucet.tsx [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+`;
+
+const LearnMoreArrow = styled(Icon)`
+  transform: rotate(180deg);
+  width: 14px;
+  height: 14px;
+`;
+
+const NoTestnetAccounts = styled(InlineMessage)`
+  margin-bottom: 20px;
```
Comment (GuillaumeRx) [inline src/features/Faucet/Faucet.tsx:68]: Use `SPACING` from `@theme` if you can to define margin/padding.

File: src/features/Faucet/Faucet.tsx [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+import { noOp } from '@utils';
+
+import { Error } from './components';
+import { possibleSolution, regenerateChallenge, requestChallenge, solveChallenge } from './helpers';
+
+// Legacy
+
+const SLabel = styled.label`
+  margin-bottom: 8px;
+  color: #333333;
```
Comment (GuillaumeRx) [inline src/features/Faucet/Faucet.tsx:39]: Use `COLORS` from `@theme` to define colors.

File: src/features/Faucet/Faucet.tsx [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+    color: var(--color-link-color);
+  }
+`;
+
+const IncorrectResponse = styled(InlineMessage)`
+  margin-top: 10px;
+`;
+
+const Description = styled.p`
+  margin-bottom: 30px;
```
Comment (GuillaumeRx) [inline src/features/Faucet/Faucet.tsx:83]: Use `SPACING` from `@theme` if you can to define margin/padding.

File: src/features/Faucet/Faucet.tsx [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+    color: var(--color-link-color);
+    text-decoration: none;
+  }
+  a:hover {
+    color: var(--color-link-color);
+  }
+`;
+
+const IncorrectResponse = styled(InlineMessage)`
+  margin-top: 10px;
```
Comment (GuillaumeRx) [inline src/features/Faucet/Faucet.tsx:79]: Use `SPACING` from `@theme` if you can to define margin/padding.

File: src/features/Faucet/Faucet.tsx [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+`;
+
+const Description = styled.p`
+  margin-bottom: 30px;
+`;
+
+const CenterText = styled.div`
+  margin-top: 30px;
```
Comment (GuillaumeRx) [inline src/features/Faucet/Faucet.tsx:87]: Use `SPACING` from `@theme` if you can to define margin/padding.

File: src/features/Faucet/Faucet.tsx [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+// Legacy
+
+const SLabel = styled.label`
+  margin-bottom: 8px;
+  color: #333333;
+  font-weight: normal;
+`;
+
+const Fieldset = styled.fieldset`
+  margin-bottom: 15px;
```
Comment (GuillaumeRx) [inline src/features/Faucet/Faucet.tsx:44]: Use `SPACING` from `@theme` if you can to define margin/padding.

File: src/features/Faucet/Faucet.tsx [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+      });
+  };
+
+  const validAccounts = accounts.filter((account) => faucetNetworks.includes(account.network.name));
+
+  const txConfig = (() => {
+    if (!('hash' in txResult)) {
+      return {} as any;
+    } else {
+      const network: any = networks.find(
```
Comment (GuillaumeRx) [inline src/features/Faucet/Faucet.tsx]: Why `any` ?

File: src/features/Faucet/Faucet.tsx [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+        value: txResult.value
+      };
+    }
+  })();
+
+  const txReceipt = (() => {
+    if (!('hash' in txResult)) {
+      return {} as any;
+    } else {
+      const network: any = networks.find(
```
Comment (GuillaumeRx) [inline src/features/Faucet/Faucet.tsx]: Why `any` ?

File: src/features/Faucet/components/Error.tsx [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+  border-radius: 2px;
+  margin-top: 40px;
+`;
+
+const ContactUsText = styled(InlineMessage)`
+  font-style: italic;
+  font-size: 14px;
+  margin: 20px;
+  a {
+    color: var(--color-link-color);
```
Comment (GuillaumeRx) [inline src/features/Faucet/components/Error.tsx:23]: Use `COLORS` from `@theme` to define colors. SCSS shouldn’t be used since we want to get rid of it

File: src/features/Faucet/components/Error.tsx [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    text-decoration: none;
+  }
+  a:hover {
+    color: var(--color-link-color);
```
Comment (GuillaumeRx) [inline src/features/Faucet/components/Error.tsx:27]: Use `COLORS` from `@theme` to define colors. SCSS shouldn’t be used since we want to get rid of it

File: src/features/Faucet/helpers.ts [hunk 16]
```diff
@@ -0,0 +1,57 @@
+import axios from 'axios';
+
+import { FAUCET_API } from '@config';
+import { StoreAccount } from '@types';
+
+const api = axios.create({
```
Comment (GuillaumeRx) [inline src/features/Faucet/helpers.ts]: Could the Faucet API logic be placed in ApiService ?

**Review by GuillaumeRx — CHANGES_REQUESTED** [review]

File: src/features/Faucet/helpers.tsx [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+    network,
+    gasPrice: txResult.gasPrice,
+    gasLimit: txResult.gasLimit,
+    nonce: txResult.nonce.toString() as ITxNonce,
+    data: txResult.data,
+    value: txResult.value
+  };
+};
+
+export const makeTxReceipt = (
```
Comment (FrederikBolding) [inline src/features/Faucet/helpers.tsx:158]: Test all helpers please!

File: src/features/RequestAssets/RequestAssets.tsx [hunk 18]
```diff
@@ -246,20 +245,20 @@ export function RequestAssets({ history }: RouteComponentProps) {
                     <Copyable
                       text={
                         isAssetToken(selectedAsset.type) &&
-                          selectedAsset.contractAddress &&
```
Comment (FrederikBolding) [inline src/features/RequestAssets/RequestAssets.tsx:249]: There is a bunch of weird formatting changes on unrelated files in your PR. Can you fix?

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 19]
```diff
@@ -160,7 +167,17 @@ const TxReceipt = ({
       const provider = new ProviderHandler(txConfig.network);
       const timestampInterval = setInterval(() => {
         getTimestampFromBlockNum(blockNumber, provider).then((transactionTimestamp) => {
-          if (sender.account && !disableAddTxToAccount) {
+          if (txReceipt && txReceipt.txType === ITxType.FAUCET) {
+            const recipientAccount = getStoreAccount(accounts)(txReceipt.to, txConfig.network.id);
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:171]: Can this ever be undefined?

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

File: src/config/constants.ts [hunk 20]
```diff
@@ -55,3 +55,7 @@ export const DEFAULT_GAP_TO_SCAN_FOR = 5;
 export const SETTINGS_FILENAME = 'MyCrypto_Settings_File';
 
 export const ASSET_DROPDOWN_SIZE_THRESHOLD = 150;
+
+// Faucet constants
+
+export const FAUCET_API = 'https://d3a0rurg7bu691.cloudfront.net';
```
[the remark this answers is not in the document]
Comment (MrLuit) [reply src/config/constants.ts]: This needs to be updated before going live, I added the production url in a comment

File: src/translations/lang/en.json [hunk 21]
```diff
@@ -346,6 +348,19 @@
...
+    "FAUCET_CLOSE": "Close",
+    "FAUCET_DESCRIPTION": "Are you a developer? Using this faucet, you can request a small amount of testnet ETH for an account on one of the supported testnets: Ropsten, Rinkeby, Kovan, and Goerli.",
+    "FAUCET_ENTER_RESPONSE": "Enter Response",
+    "FAUCET_ERROR": "If this issue keeps occurring, please [email us]($link) so we can help you and fix it.",
+    "FAUCET_NO_ACCOUNTS": "It looks like you don't have any accounts for one of the supported testnets. In order to proceed, please go ahead and [add an existing account]($link_add_account) or [create a new one]($link_create_account).",
+    "FAUCET_NOT_SURE": "Not sure what a faucet or testnet is?",
+    "FAUCET_RUNNING_LOW": "Our faucet is currently running low on this network. Please try again later.",
+    "FAUCET_SHARE": "Share on Twitter",
+    "FAUCET_SUCCESS": "Success! The testnet ETH has been sent to your wallet.",
+    "FAUCET_TWEET": "I just got testnet ETH using @MyCrypto’s faucet! Need Ropsten, Rinkeby, Kovan, or Goerli ETH? Visit mycrypto.com/faucet",
```
[the remark this answers is not in the document]
Comment (MrLuit) [reply src/translations/lang/en.json]: URL is now valid (added a constant in @config/links.ts)

File: src/components/TransactionFlow/TxReceipt.tsx (hunk 19 above)
Comment (MrLuit) [reply src/components/TransactionFlow/TxReceipt.tsx:157]: Types for the component props are defined in [src/types/transactionFlow.ts](https://github.com/MyCryptoHQ/MyCrypto/blob/afbed58548c6ffc60ebd4a7c2f3b6f65e06f319f/src/types/transactionFlow.ts#L87), it's needed because of that type definition.

```typescript
export interface IStepComponentProps extends IDefaultStepComponentProps {
  [...]
  txReceipt?: ITxReceipt;
  [...]
}
```

In reality the Faucet component will always pass a valid txReceipt prop to the TxReceipt component, not sure if any other components in the codebase ever pass `undefined` though.

File: src/services/ApiService/Faucet/constants.ts [hunk 22]
```diff
@@ -0,0 +1 @@
+export const FAUCET_API_URL = 'https://faucet.mycryptoapi.com';
```
Comment (FrederikBolding) [inline src/services/ApiService/Faucet/constants.ts:1]: Please move to `src/config/constants.ts`

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 23]
```diff
@@ -355,7 +380,7 @@ export const TxReceiptUI = ({
           }}
         />
       )}
-      {txStatus === ITxStatus.PENDING && (
+      {txStatus === ITxStatus.PENDING && txType !== ITxType.FAUCET && (
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:383]: This copy should change when its a faucet tx.

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 24]
```diff
@@ -403,6 +428,7 @@ export const TxReceiptUI = ({
         gasPrice={gasPrice}
         gasUsed={gasAmount()}
         value={txConfig.value}
+        received={txType === ITxType.FAUCET}
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:431]: Please use the new `customComponent` prop for custom UI in the TxReceipt.

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 25]
```diff
@@ -10,9 +10,18 @@ import React, {
 import { Link, RouteComponentProps, withRouter } from 'react-router-dom';
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:10]: TxReceipt does not match design
https://www.figma.com/file/BY0SWc75teEUZzws8JdgLMpy/%5BMyCrypto%5D-Web%2FApp-Final-Designs?node-id=15945%3A0

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 26]
```diff
@@ -471,38 +497,66 @@ export const TxReceiptUI = ({
           recipient={rawTransaction.to}
         />
       </div>
+      {txType === ITxType.FAUCET && (
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:500]: Can we add a prop for a custom button instead of having more conditionals in the component?

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 27]
```diff
@@ -471,38 +497,66 @@ export const TxReceiptUI = ({
...
       )}
-      {txStatus === ITxStatus.PENDING && txQueryType !== TxQueryTypes.SPEEDUP && txConfig && (
-        <Tooltip display="block" tooltip={translateRaw('SPEED_UP_TOOLTIP')}>
-          <Button
-            className="TransactionReceipt-another"
-            onClick={handleTxSpeedUpRedirect}
-            disabled={!supportsResubmit}
-          >
-            {translateRaw('SPEED_UP_TX_BTN')}
+      {txStatus === ITxStatus.PENDING &&
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:519]: Maybe add a prop flag to the conditional instead of using `txType` directly. We are trying to limit the amount of conditionals based on `txType` in the component to make it more readable.

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 28]
```diff
@@ -471,38 +497,66 @@ export const TxReceiptUI = ({
...
+              disabled={!supportsResubmit}
+            >
+              {translateRaw('CANCEL_TX_BTN')}
+            </Button>
+          </Tooltip>
+        )}
+      {txType === ITxType.FAUCET ? (
+        <Link to={ROUTE_PATHS.DASHBOARD.path}>
+          <Button secondary={true} className="TransactionReceipt-back">
+            {translate('FAUCET_CLOSE')}
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:550]: Is the difference necesary? Can we stick to one button at the bottom please?

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 29]
```diff
@@ -471,38 +497,66 @@ export const TxReceiptUI = ({
...
+              onClick={handleTxSpeedUpRedirect}
+              disabled={!supportsResubmit}
+            >
+              {translateRaw('SPEED_UP_TX_BTN')}
+            </Button>
+          </Tooltip>
+        )}
+      {txStatus === ITxStatus.PENDING &&
+        txQueryType !== TxQueryTypes.CANCEL &&
+        txType !== ITxType.FAUCET &&
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:535]: Same as above

File: src/components/TransactionFlow/TxReceiptTotals.tsx [hunk 30]
```diff
@@ -54,8 +56,17 @@ export const TxReceiptTotals = ({
     <>
       <div className="TransactionReceipt-row">
         <div className="TransactionReceipt-row-column">
-          <SIcon type="tx-send" alt="Sent" />
-          {translate('CONFIRM_TX_SENT')}
+          {received ? (
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceiptTotals.tsx:59]: Please use a `customComponent` in TxReceipt instead of this.

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 31]
```diff
@@ -471,38 +497,66 @@ export const TxReceiptUI = ({
           recipient={rawTransaction.to}
         />
       </div>
+      {txType === ITxType.FAUCET && (
+        <NewTabLink
+          href={`https://twitter.com/intent/tweet?text=${encodeURIComponent(
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:502]: Use `generateTweet`

File: src/features/Faucet/stateFactory.tsx [hunk 32]
```diff
@@ -0,0 +1,107 @@
+import { FaucetService } from '@services/ApiService/Faucet';
+import { StoreAccount } from '@types';
+import { TUseStateReducerFactory } from '@utils';
+
+import { FaucetState, ITxFaucetResult } from './types';
+
+const FaucetFactory: TUseStateReducerFactory<FaucetState> = ({ state, setState }) => {
```
Comment (FrederikBolding) [inline src/features/Faucet/stateFactory.tsx:7]: Sorry if it wasn't clear. But I would prefer we use a reducer over stateFactory.
As an example: https://github.com/MyCryptoHQ/MyCrypto/blob/master/src/features/SendAssets/SendAssets.reducer.tsx

**Review by FrederikBolding — DISMISSED** [review]

File: src/features/Faucet/helpers.spec.tsx [hunk 33]
```diff
@@ -0,0 +1,60 @@
+import { fAssets, fNetworks } from '@fixtures';
```
Comment (GuillaumeRx) [inline src/features/Faucet/helpers.spec.tsx:1]: Prefer using another method than Snapshots to test.

File: src/features/Faucet/index.tsx [hunk 34]
```diff
@@ -0,0 +1,3 @@
+import { Faucet } from './Faucet';
+
+export default Faucet;
```
Comment (GuillaumeRx) [inline src/features/Faucet/index.tsx:3]: Just `export { Faucet } from './Faucet'`

File: src/features/Faucet/types.ts [hunk 35]
```diff
@@ -0,0 +1,26 @@
+import { ITxData, ITxGasLimit, ITxGasPrice, ITxHash, TAddress } from '@types';
```
Comment (GuillaumeRx) [inline src/features/Faucet/types.ts:1]: can this go to the `types` folder ?

File: src/features/Settings/components/AddressBook.tsx [hunk 36]
```diff
@@ -240,29 +242,36 @@ export default function AddressBook({
...
-        /* eslint-disable react/jsx-key */
-        <Label>
-          <SIdenticon address={address} />
-          <SEditableText
+      ({ uuid, address, label, network, notes }: ExtendedContact, index) => {
+        const networkData = getNetworkById(network);
+        const color =
+          networkData && 'color' in networkData && networkData.color
+            ? networkData.color
+            : COLORS.LIGHT_PURPLE;
```
Comment (GuillaumeRx) [inline src/features/Settings/components/AddressBook.tsx:250]: Can this just be resumed as `networkData && networkData.color` ?

**Review by GuillaumeRx — CHANGES_REQUESTED** [review]

File: src/features/Faucet/stateFactory.tsx (hunk 32 above)
Comment (FrederikBolding) [reply src/features/Faucet/stateFactory.tsx:7]: Actually, after further thought it might be ok to use a stateFactory for this for now.

File: src/features/Faucet/types.ts (hunk 35 above)
Comment (FrederikBolding) [reply src/features/Faucet/types.ts:1]: Why? Isn't it used exclusively for this feature?

Comment (GuillaumeRx) [reply src/features/Faucet/types.ts:1]: I assumed we used some of those outside, if not then it can live here

**Review by GuillaumeRx — APPROVED** [review]

File: src/types/userAction.ts [hunk 37]
```diff
@@ -28,7 +28,8 @@ export enum ACTION_NAME {
   MYC_WINTER_START = 'myc_winter_start',
   MYC_WINTER_MID = 'myc_winter_mid',
   MYC_WINTER_END = 'myc_winter_end',
-  MYC_WINTER_BONUS = 'myc_winter_bonus'
+  MYC_WINTER_BONUS = 'myc_winter_bonus',
+  TESTNET_FAUCET = 'testnet-faucet'
```
Comment (GuillaumeRx) [inline src/types/userAction.ts:32]: Maybe use a `_` for convention ?

**Review by GuillaumeRx — CHANGES_REQUESTED** [review]

File: src/types/userAction.ts (hunk 37 above)
Comment (FrederikBolding) [reply src/types/userAction.ts:32]: Will do!

**Review by GuillaumeRx — APPROVED** [review]
